components/match2/commons/match_group_coordinates.lua

```diff
@@ -253,4 +253,36 @@ function MatchGroupCoordinates.computeCoordinates(bracket)
 	}
 end
 
+--[[
+Returns an array of matchIds in each section of a bracket. The bracket must have
+coordinates data previously computed. This returns the same sections array as
+from MatchGroupCoordinates.computeSections.
+]]
+function MatchGroupCoordinates.getSectionsFromCoordinates(bracket)
+	local sectionCount = Table.getByPathOrNil(bracket.matches, {1, 'bracketData', 'coordinates', 'sectionCount'}) or 0
+
+	local sections = Array.map(Array.range(1, sectionCount), function() return {} end)
+	for matchId in MatchGroupCoordinates.dfs(bracket) do
+		local coordinates = bracket.coordinatesByMatchId[matchId]
+		table.insert(sections[coordinates.sectionIndex], matchId)
+	end
+	return sections
+end
+
+--[[
+Returns an array of matchIds in each round of a bracket. The bracket must have
+coordinates data previously computed. This returns the same sections array as
+from MatchGroupCoordinates.computeRounds.
+]]
+function MatchGroupCoordinates.getRoundsFromCoordinates(bracket)

```

Comment:
Is this not just the same data as above but indexed differently?

Reply:
`computeRounds` returns `rounds, roundIxs`. this returns just the `rounds` part. unless I messed something up I think it's identical.

Reply:
They're both called though, so consider deduplicating?

Reply:
ok moved the common parts out to `groupMatchIdsByField`

Reply:
Ah I understand the miscommunication here, I didn't mean deduplicating the code, although that's also nice, my bad. I meant not running it twice.

Reply:
There's no natural place to put `sections` in lpdb, because it's at the match group granularity and there's no record for a match group as a whole. Also it's not useful for queries because queries would use `match2bracketdata.coordinates.sectionIndex` instead. 

I think it's better to recompute. It's really cheap because it's just a reorganization of the `sectionIndex` data already available. It's probably cheaper than storing and fetching from lpdb.

Reply:
I don't see a need to save these in LPDB. Just cache them in memory, the functions are called right after one another.

PS: Please don't mark a conversation as resolved before all participants agree.

Reply:
> PS: Please don't mark a conversation as resolved before all participants agree.

Okay

> I don't see a need to save these in LPDB. Just cache them in memory, the functions are called right after one another.

`getRoundsFromCoordinates` and `getSectionsFromCoordinates` are only called if coordinates are present in lpdb. If not then it uses `computeRounds` and `computeSections` to get the same info.

Reply:
Right, but that's what I mean, they don't both need to be called at that point

Reply:
They're not both called - it's one or the other depending on if coordinates are present in lpdb.

In `MatchGroupUtil.makeBracketFromRecords`:
```
	if firstCoordinates then
		Table.mergeInto(bracket, {
			rounds = MatchGroupCoordinates.getRoundsFromCoordinates(bracket),
			sections = MatchGroupCoordinates.getSectionsFromCoordinates(bracket),
		})
	else
		MatchGroupUtil.backfillCoordinates(bracket)
	end
```

`MatchGroupUtil.backfillCoordinates` then calls `computeRounds` and `computeSections`

Reply:
I'm very confused.

```lua
rounds = MatchGroupCoordinates.getRoundsFromCoordinates(bracket),
sections = MatchGroupCoordinates.getSectionsFromCoordinates(bracket),
```

This is calling them both?

Reply:
Oh I thought you wanted to avoid calling both `getRoundsFromCoordinates` and `computeRounds`.

Reply:
> I don't see a need to save these in LPDB. Just cache them in memory, the functions are called right after one another.

What do you mean by this then? What shouldn't be stored in LPDB?

Reply:
From what I understood from your responses, you suggested saving the result of these functions in LPDB. I don't think that is needed, since you can just cache the results in memory for the other function (or combine the functions). Think we had a bit of a miscommunication here.

Reply:
Ah ok. To clarify then - the the "reorgznied" section and rounds array (result of `getRoundsFromCoordinates` and `getSectionsFromCoordinates`) are not stored in lpdb. The rest of the coordinates data is stored in lpdb (added in #559 and #560).

Reply:
Will you be addressing my original comment?

Reply:
> Will you be addressing my original comment?

I lost track of it. what was the original concern?

Reply:
Both `getSectionsFromCoordinates` and `getRoundsFromCoordinates` iterate through the bracket when they are called after one another. That can be optimized.

Reply:
Looping through the bracket is O(n) on the size of the bracket, so I don't think there is much to optimize. Any performance gain is likely on the order of microseconds.

For me, whether something should be done in one loop or two is mostly determined by code quality considerations. Performance is almost never a factor.

Reply:
Performing work twice is a bit of a code smell, it likely means you can clean up the code a bit more. I'll allow it through this time.

Reply:
Yeah I agree that doing work twice isn't indicative of good code. However for-loops are so fast that the performance costs are negligible, so goals like having small and well-contained functions become more important.

Consider the following example:
```
local firstMatch = Array.minBy(matches, function(match) return match.date end)
local lastMatch = Array.maxBy(matches, function(match) return match.date end)
```

`Array.minBy` and `Array.maxBy` each loops over the matches array. So it's possible to speed it up by computing `firstMatch` and `lastMatch` in a single for-loop.

```
-- Option 1
local firstMatch, lastMatch = Array.minAndMaxBy(matches, function(match) return match.date end)

-- Option 2
local firstMatch, lastMatch = getFirstAndLastMatch(matches)

function getFirstAndLastMatch(matches)
   ...
end
```

I think that demonstrates the downsides of aggressively combining for-loops. In the first option, we've added `Array.minAndMaxBy`, which feels out of place in `Array` since both `Array.minBy` and `Array.maxBy` already exist. In the second option, we've implemented our own `Array.minBy/maxBy` functionality in `getFirstAndLastMatch`, which goes against principles of code reuse. 

In general, requiring for loops to be merged together makes it difficult to split large functions up. Taken to the extreme, one ends up with everything in one big function, like the current `GroupTableLeague` https://liquipedia.net/commons/Module:GroupTableLeague